### ISSUE TITLE
Periodic corrupted OFDM sympols in wifi receiver

### ISSUE THREAD

**Issue by ChengDaHaI** [bastibl/gr-ieee802-11#277]
Hi,

I am using gr-ieee802-11 (main-3.7) to collect wifi datasets on USRP X310 with two CBX-120 Dboards (with subdev:A serving as TX and subdev:B serving as RX).  

However, when I look into the BER (after Viterbi decoding), I found something wields, as shown in the figure.
![image](https://user-images.githubusercontent.com/11609265/120706577-46a85c80-c487-11eb-80ae-b4037f460311.png)
In this BER vs bit index plot, the BER is averaged over 3000 wifi packets (QAM16, 10MHz, 2.467GHz). Also, I found that each high peak (16 in total) corresponds to the (96 ) bits demodulated&decoded from one OFDM symbol. 

Motivated by this, I enumerate the packets with high errors in those high peak bits and then plot the following two figures 
![image](https://user-images.githubusercontent.com/11609265/120707228-11503e80-c488-11eb-8451-f0a1cb3c3e4b.png)
![image](https://user-images.githubusercontent.com/11609265/120707262-1b723d00-c488-11eb-916f-317c1b2143a4.png)

Note that the eqSYM is the data in equalized OFDM symbols. 

I analyze the decoded bit from the mac_out port of WIFI_PHY_Hier (with my modified WIFI_Decode_MAC to output all wifi packets, ignore the CRC results) as well as the raw IQ output of USRP receiver followed by MATLAB implemented WIFI receiver. Both give me the save BER plot. 

In addition, I have tried the wifi_loopback.grc and find there is no such high peak in BER curve. Also, I have tried the same setting one another testbed where there is UBX-160 Dboard in USRP X310 and only 3 peaks appear

![image](https://user-images.githubusercontent.com/11609265/120711292-32fff480-c48d-11eb-8ae4-2eb38a233908.png)

Thus, I guess the problem occurred in the USRP hardware side. The hardware related setting are:

-  lo_offset = 0MHz
-  uhd.tune_request(freq, rf_freq = freq - lo_offset, rf_freq_policy=uhd.tune_request.POLICY_MANUAL)
-  Mboard clock rate 200MHz
- TX normalized gain 0.7, RX normalized gain 0.1
- The distance between TX and RX antenna is about 1.5 meter
- The Dborad calibration is finished.

I would very appreciate if you could give some comments on how to debug the issue.

Best,
Hai

**Comment by bastibl** [comment]
Thanks for the great, detailed description. Is it correct that the MATLAB implementation exhibits a similar BER pattern, i.e., it's not a bug in the code? (It might yet be a common shortcomming of both implementations to deal with the hardware impariment.)

The setup with the other X310, is that also RX and TX on one device? If so, can you try to split RX and TX. There might be oscillator leakage.

Did you try offset tuning?

Are there warnings/errors/overruns/underruns indicated?

**Comment by ChengDaHaI** [comment]
> Thanks for the great, detailed description. Is it correct that the MATLAB implementation exhibits a similar BER pattern, i.e., it's not a bug in the code? (It might yet be a common shortcomming of both implementations to deal with the hardware impariment.)

- In the MATLAB implementation, I got the same BER plot (as shown). So it is not a bug in the code.

> The setup with the other X310, is that also RX and TX on one device? If so, can you try to split RX and TX. There might be oscillator leakage.

- Yes, in the two examples, TX and RX are both on one device. When you say split RX and TX, do you mean using two X310 with one serving as TX and another serving RX?

> Did you try offset tuning?

- DO you mean the lo offset tuning? If so, yes, but the default one in your example : "uhd.tune_request(freq, rf_freq = freq - lo_offset, rf_freq_policy=uhd.tune_request.POLICY_MANUAL)".

> Are there warnings/errors/overruns/underruns indicated?

- Yes, there are one warning related with URSP: "[WARNING] [MULTI_USRP] Setting IQ imbalance compensation is not possible on this device". Some underruns happens.

**Comment by bastibl** [comment]
My guess was that there is some leakage between RX and TX. What do you mean, when you said you tried LO offset tuning? The command you posted is by default issued with an lo_offset of 0. So no offset tuning. It is a parameter that can be configured through the GUI and for offset tuning should be set to a value that is larger than bandwidth/2. If you already tried this, I'm running a bit out of ideas. Maybe some other interaction between RX and TX on the same device. I'd recommend to use two devices, one acting as RX and one acting ax TX. At least, it'll help to better understand the problem.

**Comment by ChengDaHaI** [comment]
Hi Bastibl, 

Sorry for the pretty late updates... I followed your advices 

- **a: change lo_offset from 0 to 6MHz**
- **b: use TX and RX in two different USRPs**

but it seems like none of them works. Nevertheless, I have **_some results_** and can **_avoid the existense of those peaks_** by reducing the wifi packet length. In addtion, my guess about the reason for those high peaks is the **_UHD/FPGA version_**.

**TL;DR**

### Testbeds

I can access two testbed: Arena and Colosseum. Arena is a indoor testbed TX/RX transmiting over the air with fixed LOS antenna. Colosseum is a RF emulator where TX/RX connectted via cables(and FPGA which could simulate outdoor specific channel conditions). I tried lo_offset = 0 and 6Mhz on USRP X310  of both testbeds but it did not make a difference. I tried using two X310 USRP as TX and RX in Colosseum (but not in Arena due to access issues) however I get the same result with TX and RX in one X310. 

Then, what could make a difference? It could be the Dboards and UHD versions.

### Dboards

The Dboards in Colosseum are UBX-160-LP (LP, low power) while there are  CBX and UBX-160 Dboards  (difference X310) in Arena. 
- For UBX-160-LP in Colosseum, the BER is
![UBX-160-LP in Colosseum](https://user-images.githubusercontent.com/11609265/131186384-b25472f7-15ad-453f-ad5d-f6ad011c67be.png)

- For UBX-160 and CBX in Arena, the BER is the same
![Arena](https://user-images.githubusercontent.com/11609265/131186583-61477ce9-549b-4dfa-a41d-fe5ae353161d.jpg)

As you can see, the same Dboards UBX-160 (assume UBX-160 = UBX-160-LP) on Arena and Colosseum have different results while different Dboards UBX-160 and CBX on Arena have the same results. Thus, the Dboard is not the reason for peaks.

### UHD versions

- Arena UHD: UHD_3.14.0.0-0-g6875d061, FPGA Version: 35.1
- Colossem UHD: UHD_003.009.005-0-g32951af2, FPGA Version: 19.0

From UHD_usrp_probe command, the other info. are (almost) the same and thus not provided to save space. My guess is the UHD/FPGA version indeed matters. Please read the following.

### Zero-Padded Wifi Packet samples

Since those peaks are perfect periodic, I realize that the issues could be from the **wifi packets samples transmitteds at TX USRP**. Note that there is a foo packet pad2 module in the default wifi transceiver GRC, where the front padded zeros is 10k and the tail padded zeors is also 10k. 

I change the number of padded zeros and get results (On Arena X310, different Dboards give the same result ):

- For 10K padded zeros
![10K padded zeros](https://user-images.githubusercontent.com/11609265/131188458-90012b55-1043-46de-820f-132fb2094377.jpg)

- For 7280 padded zeros
![S7_UBX_1 0G_1](https://user-images.githubusercontent.com/11609265/131188495-4fc0f852-6647-4d1a-8506-f106bbb7bc3c.jpg)

- For 7K padded zeros
![S7_UBX_1 0G_2](https://user-images.githubusercontent.com/11609265/131188530-0206aa4c-45a6-4a57-9cb4-7b4ba60eb0e9.jpg)

From those results, I _guess_ that  _**the periodicly imperfect operation, like DUC (digital upper conversion) or ADC, corrupt the source wifi packet samples**_ and thus we have those high peaks in BER vs wifi bit index curves. _**Those operations (at least DUC) are related with UHD/FPGA version.**_

### Avoding those peaks

In the above figure, the PSDU length is 1024 (1000 bytes messege plus 24 bytes headers). By reducing the PSDU length to 204 ( 180 bytes messege), we get peak-free curves:
![Short_no_peak](https://user-images.githubusercontent.com/11609265/131189620-d82074c9-c08a-4aa0-8a55-5906f9454da9.png)

Note: this peak-free curve is in low-SNR and we see random small peaks all over the packeks, while the above high-peak curve is in hight-SNR and we see fixed pattern  peaks.


### Hope it helps
Alghough it is still not clear why we have those high peaks in the BER vs bit index curve, I hope this discussion could help we understanding the issue.

Best,
Hai

**Comment by bastibl** [comment]
Wow, thanks for the great report. I don't have a really good idea what could be the problem. Just some ideas:
- If possible, you could have a look at the time domain signal and see what's going on. I'm not sure if this is an issue of the signal (as received by the SDR) or some issue with the receiver.
- If it is an issue with the receiver, the first thing that comes to mind is the `d_epsilon0` and `d_er` in the frame equalizer. Maybe you could set this to 0 for testing (also commenting out their updates). https://github.com/bastibl/gr-ieee802-11/blob/maint-3.8/lib/frame_equalizer_impl.cc#L147-L148
- You could log the estimated frequency offset (it's annotated in a tag) and check if the values of the corrupted frames are different from the good frames.
